### PR TITLE
Fix RNN, StackedRNNCells with nested state_size, output_size TypeError issues

### PR DESCRIPTION
**Pull request by Ending2015a**
*repost from [tensorflow/tensorflow #50724](https://github.com/tensorflow/tensorflow/pull/50724)*
This PR fixed the following issues:
* `tf.keras.layers.RNN` generates an incorrect `state_spec` when using stacked LSTM cells.
* `tf.keras.layers.RNN` failed to build up when stacking multiple RNN cells that consume nested inputs/states and produce nested outputs.
* Add a unittest for stacked nested rnn cells

A full bug description/reproduction, please refer to [this gist](https://colab.research.google.com/drive/1WkI64errr70gCUCHU7YAz83jgx8o7VgP?usp=sharing).

Already run the unittest (keras/layers/recurrent_test.py) on my local machine.

### CODE REVIEW

File: keras/layers/recurrent.py [hunk 1]
```diff
@@ -152,6 +152,13 @@ def call(self, inputs, states, constants=None, training=None, **kwargs):
   def build(self, input_shape):
     if isinstance(input_shape, list):
       input_shape = input_shape[0]
+
+    def get_batch_input_shape(batch_size):
```
Comment (fchollet) [inline keras/layers/recurrent.py:156]: This factoring seems confusing. Consider using `get_batch_input_shape(batch_size, dim)` and calling it using `partial`

**Review by fchollet — CHANGES_REQUESTED** [review]
Thanks for the PR!

File: keras/layers/recurrent.py (hunk 1 above)
Comment (Ending2015a) [reply keras/layers/recurrent.py]: ok, I have fixed it.

**Review by fchollet — APPROVED** [review]
LGTM, thanks!
